### PR TITLE
Enable more styling options for guide_colorbar().

### PR DESCRIPTION
**Pull request by clauswilke**
Inspired by StackOverflow questions [here](https://stackoverflow.com/q/29793398/4975218) and [here:](https://stackoverflow.com/q/50070741/4975218) This patch enables better styling of the colorbar guide.

Examples:
```{r}
set.seed(123)
df <- data.frame(x <- rnorm(10),
                 y <- rnorm(10),
                 fill <- rnorm(10))

p <- ggplot(df, aes(x, y, fill = fill)) + geom_point()

# current ggplot2
p + scale_fill_gradient(low = 'white', high = 'red')
```
screen shot 2018-04-28 at 10 34 44 am

```{r}
# make frame around color bar
p + scale_fill_gradient(low = 'white', high = 'red',
                        guide = guide_colorbar(frame.colour = "black"))
```
screen shot 2018-04-28 at 10 34 54 am

```{r}
# style tick marks
p + scale_fill_gradient(low = 'white', high = 'red',
                        guide = guide_colorbar(ticks.colour = "black"))
```
screen shot 2018-04-28 at 10 35 01 am

```{r}
# style frame and tick marks at the same time, change linewidth
p + scale_fill_gradient(low = 'white', high = 'red',
                        guide = guide_colorbar(frame.colour = "black",
                                               frame.linewidth = 1,
                                               ticks.colour = "black",
                                               ticks.linewidth = 1))
```
screen shot 2018-04-28 at 10 35 08 am

This pull request will require some modifications once https://github.com/tidyverse/ggplot2/pull/2415 is integrated (or vice versa). I'm happy to take care of that.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -285,6 +285,11 @@ up correct aspect ratio, and draws a graticule.
 
 * Default themes use `rel()` to set line widths (@baptiste).
 
+### Guides
+
+* Make `guide_colorbar()` more configurable; enable styling of tick marks and
```
Comment (hadley) [inline NEWS.md]: Can you mention the argument names here please?

File: R/guide-colorbar.r [hunk 2]
```diff
@@ -109,8 +117,15 @@ guide_colourbar <- function(
   nbin = 20,
   raster = TRUE,
 
+  # frame
+  frame.colour = NULL,
+  frame.linewidth = 0.5,
+  frame.linetype = 1,
```
Comment (hadley) [inline R/guide-colorbar.r]: Have we used `linewidth` and `linetype` in other functions?

File: R/guide-colorbar.r [hunk 3]
```diff
@@ -273,6 +295,16 @@ guide_gengrob.colorbar <- function(guide, theme) {
                rectGrob(x = 0, y = by, vjust = 0, hjust = 0, width = barwidth.c, height = bh, default.units = "mm",
                         gp = gpar(col = NA, fill = guide$bar$colour))
              })
+    }
+
+  # make frame around color bar if requested (colour is not NULL)
+  if (!is.null(guide$frame.colour)) {
+    grob.bar <- grobTree(grob.bar,
```
Comment (hadley) [inline R/guide-colorbar.r]: Would you mind adhering to modern tidyverse style here?

```R

grobTree(
  grob.bar,
  rectGrob(
    width = barwidth.c, height = barheight.c, default.units = "mm",
    gp = gpar(
      col = guide$frame.colour,
      lwd = guide$frame.linewidth,
      lty = guide$frame.linetype,
      fill = NA
    )
  )
)
```

or similar

File: R/guide-colorbar.r [hunk 4]
```diff
@@ -352,7 +384,9 @@ guide_gengrob.colorbar <- function(guide, theme) {
           y1 = rep(tic_pos.c, 2)
         })
       segmentsGrob(x0 = x0, y0 = y0, x1 = x1, y1 = y1,
-                   default.units = "mm", gp = gpar(col = "white", lwd = 0.5, lineend = "butt"))
+                   default.units = "mm", gp = gpar(col = guide$ticks.colour,
```
Comment (hadley) [inline R/guide-colorbar.r]: And here

File: R/guide-colorbar.r (hunk 2 above)
Comment (clauswilke) [reply R/guide-colorbar.r]: `linetype` is a standard aesthetic, so I think that's fine. `linewidth` is usually called `size` throughout the rest of ggplot2. If you prefer `size` I'm happy to make the change, it's just that `frame.size` doesn't sound like it controls the thickness of the frame.

File: R/guide-colorbar.r (hunk 3 above)
Comment (clauswilke) [reply R/guide-colorbar.r]: Sure, will do.

File: R/guide-colorbar.r (hunk 2 above)
Comment (hadley) [reply R/guide-colorbar.r]: I don't have any particularly strong feelings here; I don't really understand how these options are supposed to fit in with the theme system.

Comment (clauswilke) [reply R/guide-colorbar.r]: Not sure. The question is whether every possible option to a guide function needs a theme equivalent. My personal opinion is no. I think there is a lot of room for improvement for making guide functions more flexible, and possibly providing different ones or frameworks for extending them, and it's impossible to add new theme components every time somebody makes some minor tweak to some guide function. Already, there are things the guide functions can do that the theme can't control, such as the `draw.ulim` option for `guide_colorbar()`.

Comment (hadley) [reply R/guide-colorbar.r]: Agreed. I'm happy to merge this PR; I just wanted to make sure we at least talked over the issues.

**Comment by clauswilke** [comment]
Just a note: I'll wait with finalizing this PR until after https://github.com/tidyverse/ggplot2/pull/2415 is integrated, since they touch the same files.

**Comment by clauswilke** [comment]
I believe this latest update addresses all concerns. I added two minimal visual tests, one for the default styling (it also tests that colors can be changed via scale_color_..., so is useful anyways I think) and one for the modified styling of both the frame and the tick lines.

**Comment by hadley** [comment]
Thank you!

**Pull request merged**